File: app/src/main/java/com/sample/demo/retrofit_advance_demo/AppClientManager.java
package com.sample.demo.retrofit_advance_demo;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import java.util.concurrent.TimeUnit;

public class AppClientManager {
    public static final int requestApiTimeOut = 10;
    private Retrofit retrofit;
    private OkHttpClient okHttpClient;
    public AppClientManager(){
        okHttpClient = createOkHttpClient();
        retrofit = createClient();
    }

    public AppClientManager(String url){
        okHttpClient = createOkHttpClient();
        retrofit = createClient(url);
    }

    private OkHttpClient createOkHttpClient(){
        HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor(new HttpLoggingInterceptor.Logger() {
            @Override
            public void log(String message) {

            }
        });
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        return new OkHttpClient.Builder()
                .addInterceptor(httpLoggingInterceptor)
                .connectTimeout(requestApiTimeOut, TimeUnit.SECONDS)
                .writeTimeout(requestApiTimeOut, TimeUnit.SECONDS)
                .readTimeout(requestApiTimeOut, TimeUnit.SECONDS)
                .build();
    }

    public Retrofit createClient(){
        return new Retrofit.Builder()
            .baseUrl(Config.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .client(okHttpClient)
            .build();
    }

    public Retrofit createClient(String url){
        return new Retrofit.Builder()
                .baseUrl(url)
                .addConverterFactory(GsonConverterFactory.create())
                .client(okHttpClient)
                .build();
    }

    public Retrofit getClient(){
        return retrofit;
    }
}

File: app/src/main/java/com/sample/demo/retrofit_advance_demo/AppClientManager.kt
package com.sample.demo.retrofit_advance_demo

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

import java.util.concurrent.TimeUnit

class AppClientManager {
    val requestApiTimeOut = 10
    var client: Retrofit
    private var okHttpClient: OkHttpClient? = null

    constructor() {
        okHttpClient = createOkHttpClient()
        client = createClient()
    }

    constructor(url: String) {
        okHttpClient = createOkHttpClient()
        client = createClient(url)
    }

    private fun createOkHttpClient(): OkHttpClient {
        val httpLoggingInterceptor = HttpLoggingInterceptor(HttpLoggingInterceptor.Logger { })
        httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
        return OkHttpClient.Builder()
            .addInterceptor(httpLoggingInterceptor)
            .connectTimeout(requestApiTimeOut.toLong(), TimeUnit.SECONDS)
            .writeTimeout(requestApiTimeOut.toLong(), TimeUnit.SECONDS)
            .readTimeout(requestApiTimeOut.toLong(), TimeUnit.SECONDS)
            .build()
    }

    fun createClient(): Retrofit {
        return Retrofit.Builder()
            .baseUrl(Config.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .client(okHttpClient!!)
            .build()
    }

    fun createClient(url: String): Retrofit {
        return Retrofit.Builder()
            .baseUrl(url)
            .addConverterFactory(GsonConverterFactory.create())
            .client(okHttpClient!!)
            .build()
    }
}
